edalize/symbiyosys.py

```diff
@@ -0,0 +1,297 @@
+import os
+import jinja2
+
+from edalize.edatool import Edatool
+
+
+class Symbiyosys(Edatool):
+
+    _description = '''SymbiYosys backend
+
+SymbiYosys is a wrapper around yosys to make it easier to do formal
+verification.
+
+Example snippet in CAPI2 format:
+
+.. code:: yaml
+
+   symbiyosys:
+     tasknames:
+       # A list of task names to pass to sby. Defaults to empty (in which case
+       # sby will run each task in the .sby file)
+       - my_proof
+
+The SymbiYosys tool expects a configuration file telling it what to do. This
+file includes a list of all the source files, together with various flags which
+we'd like to make customizable (we support vlogdefine and vlogparam).
+
+To use the Edalize wrapper, you should include exactly one source file with
+file type 'sbyConfigTemplate'. This template file is a Jinja2 template and is
+expanded to yield a ``test.sby`` in the build directory.
+
+The sby file format expects a list of input files (twice: each file needs to
+appear in a read line at the start of the ``[script]`` section, then again in
+the ``[files]`` section). To make this a bit easier, the Edalize backend
+supplies several variables and a custom filter.
+
+If you don't need to do anything particularly complicated, start your script
+section with something like this:
+
+.. code::
+
+    [script]
+    {{"-sv"|gen_reads}}
+
+This will expand to something like:
+
+.. code::
+
+    [script]
+    read -sv -Ddefine1=1 -Ddefine2=2 -Isome/include/dir file1.sv
+    read -sv -Ddefine1=1 -Ddefine2=2 -Isome/include/dir file2.sv
+    chparam -set foo 'bar' -set baz 'qux' my_toplevel
+
+
+Here, the list of files is the list of files given to Edalize in the first
+place. The include directories also come from the file list (calculated by
+Edalize's internal fileset tracking). The defines come from ``vlogdefine``
+parameters and the chparam command comes from ``vlogparam`` parameters.
+
+Similarly, write a files section like this:
+
+.. code::
+
+    [files]
+    {{files}}
+
+which will expand to
+
+.. code::
+
+    [files]
+    ../path/to/some/file1.sv
+    ../path/to/file2.sv
+
+If you need more control over the read commands at the start of the script
+section, there are some variables that can help you.
+
+  - ``{{chparam}}`` expands to the chparam command at the end of gen_reads as
+    shown above.
+
+  - ``{{flags}}`` expands to the flags (defines and includes) used for each
+    read line.
+
+  - ``{{src_files}}`` expands to a list with the basename of each source file.
```

Comment:
Can you also make the name of the toplevel available?

Reply:
Done! It's probably not quite as useful as the other bits: you generally know the top-level module of your design but (with fusesoc core files) you don't necessarily know the list of files that goes into it.